### ISSUE TITLE
Fails to install on NixOS due to vendor packages

### ISSUE THREAD

**Issue by socksy** [nodegit/nodegit#1247]
In this case, specifically I'm getting the 

    configure: error: No crypto library found!
    Try --with-libssl-prefix=PATH
     or --with-libgcrypt-prefix=PATH
     or --with-wincng on Windows

error detailed many times before — but this time with a twist. It is being called with `--with-libssl-prefix` set.
Specifically, 

    Command failed: /path/node_modules/nodegit/vendor/libssh2/configure --with-libssl-prefix=/path/node_modules/nodegit/vendor/openssl/openssl

However, the line checking is explicitly using `/usr/bin/file`, which doesn't exist on NixOS. Nothing in `/usr/bin/` does (due to the central design decision of NixOS). 

There is a [libssh2 package for NixOS already](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/libssh2/default.nix), would it be possible to use that in some capacity instead?

Alternatively, I can't see the `configure` script in the original libssh2 git repo — is it possible it was generated from configure.ac and included in this package? If so, is it possible to not do that?

**Comment by saper** [comment]
Well, you are basically on your own. You can try the package, but make sure there are development headers included.

**Comment by socksy** [comment]
OK, I have no idea how to even start doing that unfortunately. I am confused as to why you are shipping the vendor package with an autogenerated configure file rather than calling autoconfig as part of the build process. 

Probably there's a good reason, but it's because of doing this on a particular set up that the configure stage fails.

**Comment by saper** [comment]
`configure` script is meant to be portable on certain unixlike platforms and it is perfectly okay to ship it with the package. I don't know how good is the autoconf support for nixos; certainly configure cannot run under windows' cmd.exe so it isn't ultimately portable.

**Comment by socksy** [comment]
I know, but this particular one refers directly to binaries in /usr/bin,
which is not a very portable thing to do. NixOS is a Linux distribution,
the problems that none Unix systems have are different :)

**Comment by saper** [comment]
I disagree. Most probably /usr/bin/file has been standardized by POSIX. One could imagine a Linux kernel-based OS that has a quite non-UNIX userland and interface. android, maybe?

**Comment by socksy** [comment]
You are right that the existence of the "file" command is something in the POSIX standard, but the POSIX standard does not specify where this command should be implemented, or indeed which directory it should be put in. The user land of a Linux distro like NixOS is indeed all Linux and GNU. Android is quite a different matter.

I am quite perplexed why the configure file is included which makes what is previously a portable package a non-portable package — if it was a matter of being far more work to make it portable, then sure... but it looks like work was done in order to make it less portable? Is it due to a requirement of node-gyp?

I am also confused as to why you're including the source of these very common libraries in your vendor packages. Presumably it's easier than relying on them in something like [platform dependencies](https://www.npmjs.com/package/npm-platform-dependencies)? Genuine question, I don't really know the NPM world very well. 

But for instance, with OpenSSL, it's often regarded as quite dangerous to ship your own version. Obviously you can keep bumping the version up every now and again, but this is a PITA, and right now there's been [4 high to critical vulnerabilities discovered](https://www.openssl.org/news/vulnerabilities.html#y2017) since you last did it, only a few months ago.

**Comment by CMCDragonkai** [comment]
A config.log is generated when this happens. Here I have renamed it to config.txt, it reveals the errors that configure is actually meeting:

[config.txt](https://github.com/nodegit/nodegit/files/1047851/config.txt)

**Comment by CMCDragonkai** [comment]
It appears there may be a need to provide an override during the build of nodegit to sed substitute all uses of `/usr/bin` to something more general.

However there's also a configure warning about the missing script, but I don't think this is important.

**Comment by aaronjanse** [comment]
> It appears there may be a need to provide an override during the build of nodegit to use the sed/replace command to substitute all uses of /usr/bin to nix store paths.

> Most probably /usr/bin/file has been standardized by POSIX

How about we simply replace `/usr/bin/file` with `/usr/bin/env file`, as we are [supposed to do](https://stackoverflow.com/a/10383546) when referencing shell commands (e.g. `bash`).
